Customize Delta text

nicomarg opened the pull request:
Parent PR: LiveSplit/LiveSplit#2497

Added options to customize the text on the left of the component.

nicomarg committed:
Added customizable text

Added options to change the text on the left of the component.
Added an extra option to display different texts when ahead or behind the comparison.

wooferzfg commented on src/LiveSplit.Delta/UI/Components/DeltaComponent.cs:
```diff
@@ -156,6 +148,19 @@ public void Update(IInvalidator invalidator, LiveSplitState state, float width,
+                InternalComponent.AlternateNameText.Clear();
+                InternalComponent.AlternateNameText.Add(CompositeComparisons.GetShortComparisonName(comparison));
+            }
+            InternalComponent.LongestString = text;
```
This needs to be the longest possible string out of all the values it can switch between, otherwise the horizontal layout will resize when the text changes.

wooferzfg replied:
Specifically, I think it needs to be the longer of `Settings.CustomText` and `Settings.CustomTextAhead` when there's an override

nicomarg replied:
I can write it like that, though the old behaviour (when set to Current Comparison) was to update `LongestString` to the (current) long comparison name at each update, not the longest it can be without changing the settings.
If this variable is meant to be changed while running (previously it was at each current comparison change), I don't see the point of telling the renderer to make space for something longer than we need it to display.
If it's not meant to be changed regularly, would it be better to set it to the longest comparison name when set to current comparison ?
Or is it a kind of "middle ground", meaning changing comparison is meant to be relatively rare and at that point changing layout spacing isn't a big deal, while being ahead/behind changes often enough to warrant always making enough space for both texts ?

wooferzfg replied:
The issue with your current code is that the text changes between ahead vs. behind, but `LongestString` doesn't account for this. That's the only case that's problematic in my opinion.

Changing the comparison or the settings isn't something we need to account for here
(ideally changing the comparison wouldn't cause a resize either, but since that's already an issue outside of this PR, no need to handle it)

wooferzfg commented on src/LiveSplit.Delta/UI/Components/DeltaComponent.cs:
```diff
@@ -156,6 +148,19 @@ public void Update(IInvalidator invalidator, LiveSplitState state, float width,
+            if (InternalComponent.InformationName != text)
+            {
+                InternalComponent.AlternateNameText.Clear();
+                InternalComponent.AlternateNameText.Add(CompositeComparisons.GetShortComparisonName(comparison));
```
Are you sure you want this? This will show the short comparison name instead of the custom text when the custom text starts to get cut off.

nicomarg replied:
If I'm understanding correctly, it will only display the short comparison name if the custom text is deemed too long. Though when thinking about it again; it seems counter-intuitive enough to change this behaviour.

wooferzfg commented on src/LiveSplit.Delta/UI/Components/DeltaComponent.cs:
```diff
@@ -156,6 +148,19 @@ public void Update(IInvalidator invalidator, LiveSplitState state, float width,
+            var text = comparisonName;
+            if (Settings.OverrideText)
+            {
+                text = Settings.DifferentialText && InternalComponent.TimeValue < TimeSpan.Zero ? Settings.TextAhead : Settings.TextBehind;
```
Does this mean you're using `TextBehind` when the runner is ahead if `DifferentialText` is off? That's pretty confusing - maybe change the naming and/or add some comments to explain this behavior?

nicomarg replied:
That's the intended behavior (only the settings box for `TextBehind` is enabled when `DifferentialText` is off). I'll change the name to better signify that this variable acts as the overridden text when it is constant, or the text when the runner is behind when there's a different one for when the runner is ahead.

wooferzfg commented on src/LiveSplit.Delta/UI/Components/DeltaSettings.cs:
```diff
@@ -153,12 +171,33 @@ private int CreateSettingsNode(XmlDocument document, XmlElement parent)
+        {
+            chkDifferentialText.Enabled = label2.Enabled = txtBehind.Enabled = chkOverrideText.Checked;
+            label3.Enabled = txtAhead.Enabled = chkOverrideText.Checked && chkDifferentialText.Checked;
+            label2.Text = chkDifferentialText.Checked ? "Text when behind:" : "Text:";
```
Please rename the labels as well, to make it clearer what they correspond to

nicomarg committed:
No AlternateText for custom text

Fixed the counter-intuitive behaviour of still providing the short comparison name as an alternate text when using custom text.
Also fixes LongestString not necessarily being the longest alternative

nicomarg committed:
Better variable naming

Renamed labels used in the code.
Renamed TextBehind/TextAhead to CustomText/CustomTextAhead to better express the fact that the first one is used as the sole text in constant text mode and as the text when behind otherwise.

nicomarg commented:
Last 2 commits fix issues pointd earlier.
For anyone willing to read the Designer diff, idk why Designer changed some refs to class members from `member` to `this.member` when it is declared in the other file describing the partial class.

wooferzfg committed:
Clean up settings

nicomarg committed:
Fix LongestString changing with the text

In differential mode, LongestString is now always the longest of the two strings, to avoid causing resizings.
